### ISSUE TITLE
is it safe to read/write a Buffer in different threads

### ISSUE THREAD

**Issue by bruceauyeung** [node-ffi-napi/node-ffi-napi#175]
scenario:
~~~
var astruct = new MyStruct(); // implemented with ref-struct
var aref =astruct .ref();
const ret = myFunc.async(aref,(err,res)=>{});
readMyBuffer(aref);
~~~
the memory block (according to by `astruct `) is modified by `myFunc`, IIRC this runs in random libuv thread my this case.
`readMyBuffer` runs in main event loop thread.
is this scenario safe ?
if not, how can i archieve my purpose?
any suggestions are appreciated!

@addaleax @TooTallNate

**Comment by addaleax** [comment]
@bruceauyeung Friendly reminder that pinging people who are not involved in project maintenance again will get you blocked immediately.

You could use `readMyBuffer` in the callback to make this safe, i.e. `myFunc.async(aref,(err,res)=>{ readMyBuffer(aref); });`.

**Comment by bruceauyeung** [comment]
> @bruceauyeung Friendly reminder that pinging people who are not involved in project maintenance again will get you blocked immediately.

sorry for that , i didn't even notice that he is not a maintainer anymore

>You could use readMyBuffer in the callback to make this safe, i.e. myFunc.async(aref,(err,res)=>{ readMyBuffer(aref); });

in your code snippet `readMyBuffer` runs also in main event loop thread just like `readMyBuffer` in my code snippet, `myFunc ` modify struct data by pointer in a different libuv thread without any data sync , why is your code snippet safe ? what's the difference between yours and mine ?

@addaleax

**Comment by addaleax** [comment]
The difference is that code in the callback only runs after `myFunc` has fully finished running on that other thread.

**Comment by bruceauyeung** [comment]
do you mean after `myFunc`  has fully finished running on that other thread, `libuv`/`ffi` ensures any modifications on `astruct ` pointered by `aref` are synced so that main event loop thread can see that?

so the following code snippet is safe now, right ?
~~~
function asyncInvoke<T>(ffiFn: any, ...args: any[]): Promise<T> {
  return new Promise<T>(resolve => {
    ffiFn.async(...args, (err: Error, res: T) => {
      if (err) {
        logger.error(err);
      }
      resolve(res);
    });
  });
}

var astruct = new MyStruct(); // implemented with ref-struct
var aref =astruct .ref();
const ret = await asyncInvoke(myFunc, aref);
readMyBuffer(aref);
~~~

**Comment by addaleax** [comment]
Yes, if there is a clear before-after relationship between memory accesses then it’s fine to access the memory from multiple threads.

**Comment by bruceauyeung** [comment]
> 
> 
> Yes, if there is a clear before-after relationship between memory accesses then it’s fine to access the memory from multiple threads.

many thanks  for your clarification! 
i don't play c/c++ for a long time.  if i understand correctly in Java/c++ , only variables marked as `volatile` in this kind of `one thread write, other threads read` are safe, in theory.

@addaleax

**Comment by addaleax** [comment]
No, `volatile` has nothing to do with this at all. The fact that there is a before-after relationship between the function call and the callback already is proper data synchronization.
